src/transformers/models/bartpho/tokenization_bartpho.py

```diff
@@ -151,27 +147,30 @@ def __init__(
             **kwargs,
         )
 
-        self.vocab_file = vocab_file
-        self.monolingual_vocab_file = monolingual_vocab_file
         self.sp_model = spm.SentencePieceProcessor(**self.sp_model_kwargs)
         self.sp_model.Load(str(vocab_file))
+        self.vocab_file = vocab_file
+
+        # Original fairseq vocab and spm vocab must be "aligned":
+        # Vocab    |    0    |    1    |   2    |    3    |  4  |  5  |  6  |   7   |   8   |  9
+        # -------- | ------- | ------- | ------ | ------- | --- | --- | --- | ----- | ----- | ----
+        # fairseq  | '<s>'   | '<pad>' | '</s>' | '<unk>' | ',' | '.' | '▁' | 's'   | '▁de' | '-'
+        # spm      | '<unk>' | '<s>'   | '</s>' | ','     | '.' | '▁' | 's' | '▁de' | '-'   | '▁a'
 
-        # Load the reduced vocab
-
-        # Keep order of special tokens for backward compatibility
-        self.fairseq_tokens_to_ids = {}
-        cnt = 0
-        for token in [bos_token, pad_token, eos_token, unk_token, sep_token, cls_token]:
-            if str(token) not in self.fairseq_tokens_to_ids:
-                self.fairseq_tokens_to_ids[str(token)] = cnt
-                cnt += 1
-        with open(monolingual_vocab_file, "r", encoding="utf-8") as f:
-            for line in f.readlines():
-                token = line.strip().split()[0]
-                self.fairseq_tokens_to_ids[token] = len(self.fairseq_tokens_to_ids)
-        if str(mask_token) not in self.fairseq_tokens_to_ids:
-            self.fairseq_tokens_to_ids[str(mask_token)] = len(self.fairseq_tokens_to_ids)
```

Comment:
Thanks for the explanations! 

To make sure I understand everything, why it was not satisfactory to use `MBartTokenizer`?

```python
model_name = "vinai/bartpho-syllable"
tokenizer_mbart = MBartTokenizer.from_pretrained("vinai/bartpho-syllable")
```

Reply:
A `fairseq` BARTpho sequence has the following format for a single sentence X: `<s> X </s>`, while MBartTokenizer will produce `X </s> src_lang_code`.

Reply:
I am wondering whether you would have any further concern regarding this PR ? @SaulLu 
(cc @LysandreJik , @sgugger , @patil-suraj and @patrickvonplaten )

Reply:
Thank you for the previous explanations.

For my part, I think my main feedback is [this one](https://github.com/huggingface/transformers/pull/17254#discussion_r876742460)

From my point of view, it may be possible to make a fast tokenizer but it will probably take some hacking to get the same tokenization strategy.

Reply:
@SaulLu Thank you very much for feedback. 
My response to your feedback is at [here](https://github.com/huggingface/transformers/pull/17254#discussion_r876864300)

Reply:
@SaulLu I revised my code so that now the fast tokenizer has the same tokenization strategy as of the slow original one, doing some hacking in the function `_convert_encoding`. 

In particular, I extend the monolingual vocab with all other tokens from the multilingual vocab. 

```python
class BartphoConverter(SpmConverter):
    def vocab(self, proto):

        dict = {piece.piece: piece.score for piece in proto.pieces[3:]}

        # Original monolingual vocab where `mask_token_id = len(vocab) - 1` 
        vocab = [
            (k, dict[k] if k in dict.keys() else 0.0) for k, v in self.original_tokenizer.fairseq_tokens_to_ids.items()
        ]

       # Extend the monolingual vocab with all remaining tokens from the pretrained XLM-R sentencepiece model.
        for piece in proto.pieces[3:]:
            if piece.piece not in self.original_tokenizer.fairseq_tokens_to_ids.keys():
                vocab += [(piece.piece, piece.score)]

        return vocab
```

The trick is to map the token with its id that is larger than `mask_token_id` into `unk_token_id` in the function `_convert_encoding` of the fast tokenizer: https://github.com/datquocnguyen/transformers/blob/1b22570797d8e70de0b6dabccf8c9c4ececff3b9/src/transformers/models/bartpho/tokenization_bartpho_fast.py#L175 

```python
        for e in encodings:
            #encoding_dict["input_ids"].append(e.ids)
            ids = [id if id <= self.mask_token_id else self.unk_token_id for id in e.ids]
            encoding_dict["input_ids"].append(ids)
```
cc: @sgugger FYI, I keep the original slow tokenizer intact. Changes now do not break the existing code.

Reply:
@SaulLu I revised my code so that now the fast tokenizer has the same tokenization strategy as of the slow original one, doing some hacking in the function `_convert_encoding`. 

In particular, I extend the Vietnamese monolingual vocab with all other tokens from the multilingual vocab. So now I have a variant of the multilingual vocab, where the first half stores all tokens from the Vietnamese monolingual vocab and the second half stores the rest, separated by the "mask_token". 

```python
class BartphoConverter(SpmConverter):
    def vocab(self, proto):

        dict = {piece.piece: piece.score for piece in proto.pieces[3:]}

        # Original monolingual vocab where `mask_token_id = len(vocab) - 1` 
        vocab = [
            (k, dict[k] if k in dict.keys() else 0.0) for k, v in self.original_tokenizer.fairseq_tokens_to_ids.items()
        ]

       # Extend the monolingual vocab with all remaining tokens from the pretrained XLM-R sentencepiece model.
        for piece in proto.pieces[3:]:
            if piece.piece not in self.original_tokenizer.fairseq_tokens_to_ids.keys():
                vocab += [(piece.piece, piece.score)]

        return vocab
```

The trick is to map the token with its id that is larger than `mask_token_id` into `unk_token_id` in the function `_convert_encoding` of the fast tokenizer: https://github.com/datquocnguyen/transformers/blob/1b22570797d8e70de0b6dabccf8c9c4ececff3b9/src/transformers/models/bartpho/tokenization_bartpho_fast.py#L175 

```python
        for e in encodings:
            #encoding_dict["input_ids"].append(e.ids)
            ids = [id if id <= self.mask_token_id else self.unk_token_id for id in e.ids]
            encoding_dict["input_ids"].append(ids)
```
cc: @sgugger FYI, I keep the original slow tokenizer intact. Changes now do not break the existing code.